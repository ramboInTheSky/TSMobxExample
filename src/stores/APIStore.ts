import { observable, action, toJS } from 'mobx'
import axios from 'axios'

import { getErrorCode, toSnakeCase, toCamelCase } from '../utils'

export { observable, action }

export class Store<T> {
	private url: string
	private fresh = false

	@observable protected data: Array<T> = []

	@observable protected detail: T | Partial<T> = {}

	@observable protected count: number = 0

	@observable protected loading: boolean = false

	@observable public error?: string = undefined

	constructor(uri: string) {
		this.url = uri
	}
	// public API
	// public getItems = (): Array<T> => {
	// 	return toJS(this.data) as Array<T>
	// }

	public get items(): Array<T> {
		return toJS(this.data) as Array<T>
	}

	public get item(): T {
		return toJS(this.detail) as T
	}

	// public getItem = (): T => {
	// 	return toJS(this.detail) as T
	// }

	public get errorMessage(): string {
		return this.error as string
	}

	public get isLoading(): boolean {
		return this.loading
	}

	public get isError(): boolean {
		return !!this.error
	}

	public isFresh = () => this.fresh
	// end public API

	protected apiEnded = (e?: any, customMessage?: string) => {
		this.loading = false
		this.error = e ? customMessage || getErrorCode(e) : undefined
		this.fresh = true
		const self = this
		setTimeout(function () {
			self.fresh = false
		})
	}

	@action
	async getList(url?: string, filter?: string) {
		if (!this.url && !url) return
		try {
			this.loading = true
			let res: any
			res = await axios.get(url || this.url + '?filter=' + filter)
			this.apiEnded()
			this.count = res.data && res.data.count ? res.data.count : 0
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			this.count = 0
			return []
		}
	}

	@action
	async getDetail(item: string) {
		try {
			this.loading = true
			let res: any
			res = await axios.get(`${this.url}/${item}`)
			this.apiEnded()
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			return {}
		}
	}

	@action
	async save(item: Partial<T>, id?: string, url?: string) {
		try {
			this.loading = true
			let res: any
			if (id) {
				res = await axios.put(`${url || this.url}/${id}`, toSnakeCase(item))
			} else {
				res = await axios.post(url || this.url, toSnakeCase(item))
			}
			this.apiEnded()
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			return {}
		}
	}
}
