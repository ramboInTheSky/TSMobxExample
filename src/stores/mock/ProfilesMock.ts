import { Profile } from '../../models/Profile'

export class ProfilesMock {
    public item: Profile = {
        id: '1aa657bc-d556-4643-a68f-af46f1bdba89',
        firstName: 'Clark',
        lastName: 'Kent',
        company: 'test 123',
        mobileNumber: '01234567890',
        emailAddress: 'dsg@dsf.com',
        notes: 'fdsdfsfs',
        created: '2019-04-16T12:03:37.493156',
        visits: [
            {
                id: 'd3aa88e2-c754-41e0-8ba6-4198a34aa0a2',
                lockGroups: ['Default', 'Boiler', '1.11'],
                fromDate: '2015-03-23T11:00:00Z',
                toDate: '2015-11-25T12:00:00Z',
                tags: ['123123', '234234', '345345'],
                archetypeIds: [],
            },
        ],
    }
    public detail = this.item
    public items: Profile[] = [
        {
            id: '1aa657bc-d556-4643-a68f-af46f1bdba89',
            firstName: 'Clark',
            lastName: 'Kent',
            company: 'test 123',
            mobileNumber: '01234567890',
            emailAddress: 'dsg@dsf.com',
            notes: 'fdsdfsfs',
            created: '2019-04-16T12:03:37.493156',
            visits: [
                {
                    id: 'd3aa88e2-c754-41e0-8ba6-4198a34aa0a2',
                    lockGroups: ['Default', 'Boiler', '1.11'],
                    fromDate: '2015-03-23T11:00:00Z',
                    toDate: '2015-11-25T12:00:00Z',
                    tags: ['123123', '234234', '345345'],
                    archetypeIds: [],
                },
            ],
        },
    ]
    public data = this.items
    public isLoading: boolean = false
    public isError: boolean = false
    public getDetail: Function = jest.fn()
    public getList: Function = jest.fn()
    public toggleEditing: Function = jest.fn()
}
