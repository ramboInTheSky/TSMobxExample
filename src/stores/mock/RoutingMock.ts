import { Routing } from '../Routing';

export class RoutingMock extends Routing{
	goToPage:any = jest.fn()
	goBackToPage:any = jest.fn()
	push:any = jest.fn()
	replace:any = jest.fn()
	go:any = jest.fn()
	goBack:any = jest.fn()
	goForward:any = jest.fn()
	breadcrumb: Array<any> = []
	location: any = {}
	history: any = {}
	getBreadcrumb: any = jest.fn()
	resizeBreadcrumb: any = jest.fn()
	splitThings: any = jest.fn()
	rebuildBreadcrumb: any = jest.fn()
}