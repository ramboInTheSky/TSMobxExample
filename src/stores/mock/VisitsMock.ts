import { Archetype, Visit } from '../../models/Visit';
import { Tag } from '../../models/Tag';

export class VisitsMock{
    archetypes:{
		primary: Array<Archetype>,
		secondary: Array<Archetype>
	} = {
			primary: [],
			secondary: []
		}
		getArchetypes: Function = jest.fn()
		detail: Visit = new Visit()
		item = ()=> this.detail
		addTag = (tag: Tag, profileId: string, visitId: string) => {this.detail.tags.push(tag.tagNumber!)}
		clearErrors = jest.fn()
		validationError?: string = undefined
		simulateValidationError = () => this.validationError = 'this is an error'
}