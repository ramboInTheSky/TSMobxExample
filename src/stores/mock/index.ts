export * from './RoutingMock'
export * from './ProfilesMock'
export * from './VisitsMock'

export const mockDateConstructor = () => {
    const now = new Date('2020')
    /*eslint no-global-assign:off*/
    Date = class extends Date {
        constructor() {
            super()
            return now
        }
    } as DateConstructor
}