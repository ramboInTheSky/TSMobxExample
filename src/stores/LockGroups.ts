import { observable, action } from 'mobx'
import { Store } from './APIStore'
import { LockGroup } from '../models/LockGroups'
import urls from '../constants/urls'

const sortingFn = (a: LockGroup, b: LockGroup) => {
	const first = a.name
	const last = b.name
	if (first > last) return -1
	if (last > first) return 1
	return 0
}

export class LockGroupStore extends Store<LockGroup> {
	@observable
	public selectedLockGroups: Array<LockGroup> = []

	constructor() {
		super(urls.lockGroups)
	}

	async getList(filter?: string) {
		const data = await super.getList(undefined, filter)
		this.data =
			data && data.items ? data.items : []
	}

	@action 
	setList(newList: Array<LockGroup>){
		this.data = newList
	}

}
