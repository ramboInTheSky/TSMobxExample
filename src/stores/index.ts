// first party
import { ProfileStore } from './Profiles'
import { Routing } from './Routing'
import { VisitStore } from './Visits'
import { LockGroupStore } from './LockGroups'

// re-exports
export * from './Profiles'
export * from './Routing'
export * from './Visits'
export * from './LockGroups'

export interface Stores {
	profiles: ProfileStore
	routing: Routing
	visits: VisitStore
	lockGroups: LockGroupStore
}

type StoreNamesGuard = keyof Stores

export class StoreNames {
	static profiles: StoreNamesGuard = 'profiles'
	static routing: StoreNamesGuard = 'routing'
	static visits: StoreNamesGuard = 'visits'
	static lockGroups: StoreNamesGuard = 'lockGroups'
}

const localStores: Partial<Stores> = {
	profiles: new ProfileStore(),
	routing: new Routing(),
	visits: new VisitStore(),
	lockGroups: new LockGroupStore(),
}

//test comment
export const stores = { ...localStores }
