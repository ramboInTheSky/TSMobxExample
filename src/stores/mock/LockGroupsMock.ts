import { LockGroup } from '../../models/LockGroups';

export class LockGroupMock {
    public detail: LockGroup = new LockGroup()
    public data = [
        {
            "id": "123-123-123-123",
            "name": "1.11",
            "type": "room 11",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.12",
            "type": "room 12",
            "notes": "this is the best room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.13",
            "type": "room 13",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.14",
            "type": "room 14",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.15",
            "type": "room 15",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.16",
            "type": "room 16",
            "notes": "this is just a room"
        },
        {
            "id": "123-123-123-123",
            "name": "1.17",
            "type": "room 17",
            "notes": "this is an unlucky room"
        }
    ]
    public getList = jest.fn()
    public items = []
    public item = () => this.detail
}