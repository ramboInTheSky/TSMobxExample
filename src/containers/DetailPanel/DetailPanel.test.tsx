import React from 'react'
import { shallow } from 'enzyme'
import { createShallow, createMount } from '@material-ui/core/test-utils'
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { ProfileStore, } from '../../stores'
configure({ adapter: new Adapter() })

import { DetailPanel, DetailPanelConnectedProps } from '.'
import { RoutingMock, VisitsMock } from '../../stores/mock';

describe('<DetailPanel />', () => {
	let props: DetailPanelConnectedProps
	let mount: any
	beforeEach(() => {
		props = {
			callback: jest.fn(),
			profiles: new ProfileStore(),
			routing: new RoutingMock() as any,
			visits: new VisitsMock() as any
		}
		mount = createShallow()
	})
	it('renders correctly with no data', () => {
		expect(shallow(<DetailPanel {...props} />)).toMatchSnapshot()
	})

	it('renders correctly with data', () => {
		const newProps = {
			...props,
			data: {
				profileId: 'wert34-324c523-2c356-c34c523-4c23456v23',
				firstName: 'Bruce',
				lastName: 'Wayne',
				company: 'Wayne Enterprises',
				mobileNumber: '555-123-123',
				emailAddress: 'info@waine-enterprises.com',
				notes: 'Gotham is safe',
				created: '2014-03-25T12:00:00Z',
				type: undefined,
				lockGroups: ['Default', 'Boiler', '1.11'],
				visits: [
					{
						id: 'q3w4c5-w34cr-w34cw34-5cw3-4cw345-3cw34t',
						from: '2015-03-25T12:00:00Z',
						to: '2015-03-25T12:00:00Z',
						tags: ['123123', '234234', '345345'],
					},
					{
						id: 'vvv4c5-w34cr-w34cw34-5cbb-4cw345-3cw311',
						from: '20190401T10:00:10.000Z',
						to: '20190501T10:00:10.000Z',
						tags: ['123123', '234234', '345345'],
					},
				],
			},
		}
		expect(shallow(<DetailPanel {...newProps} />)).toMatchSnapshot()
	})

	it('sets edit mode on edit click and display correctly', () => {
		const el = mount(<DetailPanel {...props} />)
		el
			.dive()
			.find('.detail-panel-edit-button')
			.simulate('click', { preventDefault: () => { }, target: { name: 'input', value: 'value' } })
		el.update()
		expect(el).toMatchSnapshot()
	})

	it('correctly goes back to non-edit mode', () => {
		const el = mount(<DetailPanel {...props} />)
		el
			.dive()
			.find('.detail-panel-edit-button')
			.simulate('click', { preventDefault: () => { }, target: { name: 'input', value: 'value' } })
		el.update()
		el
			.dive()
			.find('.cancel-edit-button')
			.simulate('click', { preventDefault: () => { }, target: { name: 'input', value: 'value' } })
		el.update()
		expect(el).toMatchSnapshot()
	})

	it('calls the callback on save', () => {
		const el = mount(<DetailPanel {...props} />)
		el
			.dive()
			.find('.detail-panel-edit-button')
			.simulate('click', { preventDefault: () => { }, target: { name: 'input', value: 'value' } })
		el.update()
		el
		.dive()
		.find('.detail-panel-save-button')
		.simulate('click', { preventDefault: () => { }, target: { name: 'input', value: 'value' } })
		expect(props.callback).toHaveBeenCalled()
	})

	it('routes out on add visit', () => {
		const el = mount(<DetailPanel {...props} />)
		el
		.dive()
		.find('.detail-panel-add-visit-button')
		.simulate('click', { preventDefault: () => { }, target: { name: 'input', value: 'value' } })
		expect(props.routing.goToPage).toHaveBeenCalled()
	})
})
