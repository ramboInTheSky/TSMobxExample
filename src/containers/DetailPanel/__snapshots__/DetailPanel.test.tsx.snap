// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailPanel /> correctly goes back to non-edit mode 1`] = `
<DetailPanel
  callback={[MockFunction]}
  profiles={
    ProfileStore {
      "apiEnded": [Function],
      "count": 0,
      "data": Array [],
      "detail": Object {},
      "edit": true,
      "error": undefined,
      "fresh": false,
      "isFresh": [Function],
      "loading": false,
      "url": "https://csi-dev.thecollective.com/ac-api/profile",
    }
  }
  routing={
    RoutingMock {
      "breadcrumb": Array [],
      "getBreadcrumb": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goBackToPage": [MockFunction],
      "goForward": [MockFunction],
      "goToPage": [MockFunction] {
        "calls": Array [
          Array [
            "/",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "history": Object {},
      "location": Object {},
      "push": [MockFunction],
      "replace": [MockFunction],
      "splitThings": [MockFunction],
    }
  }
  visits={
    VisitsMock {
      "addTag": [Function],
      "archetypes": Object {
        "primary": Array [],
        "secondary": Array [],
      },
      "clearErrors": [MockFunction],
      "detail": Visit {
        "archetypeIds": Array [],
        "fromDate": "",
        "id": "",
        "lockGroups": Array [],
        "tags": Array [],
        "toDate": "",
      },
      "getArchetypes": [MockFunction],
      "item": [Function],
      "simulateValidationError": [Function],
      "validationError": undefined,
    }
  }
/>
`;

exports[`<DetailPanel /> renders correctly with data 1`] = `
<DetailPanel
  callback={[MockFunction]}
  data={
    Object {
      "company": "Wayne Enterprises",
      "created": "2014-03-25T12:00:00Z",
      "emailAddress": "info@waine-enterprises.com",
      "firstName": "Bruce",
      "lastName": "Wayne",
      "lockGroups": Array [
        "Default",
        "Boiler",
        "1.11",
      ],
      "mobileNumber": "555-123-123",
      "notes": "Gotham is safe",
      "profileId": "wert34-324c523-2c356-c34c523-4c23456v23",
      "type": undefined,
      "visits": Array [
        Object {
          "from": "2015-03-25T12:00:00Z",
          "id": "q3w4c5-w34cr-w34cw34-5cw3-4cw345-3cw34t",
          "tags": Array [
            "123123",
            "234234",
            "345345",
          ],
          "to": "2015-03-25T12:00:00Z",
        },
        Object {
          "from": "20190401T10:00:10.000Z",
          "id": "vvv4c5-w34cr-w34cw34-5cbb-4cw345-3cw311",
          "tags": Array [
            "123123",
            "234234",
            "345345",
          ],
          "to": "20190501T10:00:10.000Z",
        },
      ],
    }
  }
  profiles={
    ProfileStore {
      "apiEnded": [Function],
      "fresh": false,
      "isFresh": [Function],
      "url": "https://csi-dev.thecollective.com/ac-api/profile",
    }
  }
  routing={
    RoutingMock {
      "breadcrumb": Array [],
      "getBreadcrumb": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goBackToPage": [MockFunction],
      "goForward": [MockFunction],
      "goToPage": [MockFunction],
      "history": Object {},
      "location": Object {},
      "push": [MockFunction],
      "replace": [MockFunction],
      "splitThings": [MockFunction],
    }
  }
  visits={
    VisitsMock {
      "addTag": [Function],
      "archetypes": Object {
        "primary": Array [],
        "secondary": Array [],
      },
      "clearErrors": [MockFunction],
      "detail": Visit {
        "archetypeIds": Array [],
        "fromDate": "",
        "id": "",
        "lockGroups": Array [],
        "tags": Array [],
        "toDate": "",
      },
      "getArchetypes": [MockFunction],
      "item": [Function],
      "simulateValidationError": [Function],
      "validationError": undefined,
    }
  }
/>
`;

exports[`<DetailPanel /> renders correctly with no data 1`] = `
<DetailPanel
  callback={[MockFunction]}
  profiles={
    ProfileStore {
      "apiEnded": [Function],
      "fresh": false,
      "isFresh": [Function],
      "url": "https://csi-dev.thecollective.com/ac-api/profile",
    }
  }
  routing={
    RoutingMock {
      "breadcrumb": Array [],
      "getBreadcrumb": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goBackToPage": [MockFunction],
      "goForward": [MockFunction],
      "goToPage": [MockFunction],
      "history": Object {},
      "location": Object {},
      "push": [MockFunction],
      "replace": [MockFunction],
      "splitThings": [MockFunction],
    }
  }
  visits={
    VisitsMock {
      "addTag": [Function],
      "archetypes": Object {
        "primary": Array [],
        "secondary": Array [],
      },
      "clearErrors": [MockFunction],
      "detail": Visit {
        "archetypeIds": Array [],
        "fromDate": "",
        "id": "",
        "lockGroups": Array [],
        "tags": Array [],
        "toDate": "",
      },
      "getArchetypes": [MockFunction],
      "item": [Function],
      "simulateValidationError": [Function],
      "validationError": undefined,
    }
  }
/>
`;

exports[`<DetailPanel /> sets edit mode on edit click and display correctly 1`] = `
<DetailPanel
  callback={[MockFunction]}
  profiles={
    ProfileStore {
      "apiEnded": [Function],
      "count": 0,
      "data": Array [],
      "detail": Object {},
      "edit": true,
      "error": undefined,
      "fresh": false,
      "isFresh": [Function],
      "loading": false,
      "url": "https://csi-dev.thecollective.com/ac-api/profile",
    }
  }
  routing={
    RoutingMock {
      "breadcrumb": Array [],
      "getBreadcrumb": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goBackToPage": [MockFunction],
      "goForward": [MockFunction],
      "goToPage": [MockFunction],
      "history": Object {},
      "location": Object {},
      "push": [MockFunction],
      "replace": [MockFunction],
      "splitThings": [MockFunction],
    }
  }
  visits={
    VisitsMock {
      "addTag": [Function],
      "archetypes": Object {
        "primary": Array [],
        "secondary": Array [],
      },
      "clearErrors": [MockFunction],
      "detail": Visit {
        "archetypeIds": Array [],
        "fromDate": "",
        "id": "",
        "lockGroups": Array [],
        "tags": Array [],
        "toDate": "",
      },
      "getArchetypes": [MockFunction],
      "item": [Function],
      "simulateValidationError": [Function],
      "validationError": undefined,
    }
  }
/>
`;
